### ISSUE TITLE
Use of PDist loss for Stacked-What-Where-AutoEncoder

### ISSUE THREAD

**Issue by hawescm1** [vlfeat/matconvnet#946]
I am attempting to use a DAG network in MatConvNet to implement the Stacked-What-Where-Autoencoder (SWWAE) of:

Zhao, J., M. Mathieu, R. Goroshin, and Y. LeCunn. "Stacked What-Where Autoencoders," ICLR, 2016.

This architecture fuses a traditional layered CNN architecture for classification, with an autoencoder, where the encoder portion of the autoencoder (an encoder and a decoder) shares all but the final fully connected or global average pooling layers of the traditional CNN. This architecture enables fully supervised training (all labeled training data), fully unsupervised training (all unlabeled data), and semi-supervised training (use of both labeled and unlabeled data during training).

If one trains only with labeled data, then the autoencoder portion of the architecture serves to regularize the training with reconstruction losses. In this SWWAE architecture, the input images are compared to the output of the autoencoder's decoder output, via an L2 loss. Similarly, various intermediate feature maps in the autoencoder's encoder are compared to their counterpart feature maps in the autoencoder's decoder. In both cases, the input images or encoder feature maps are the 'targets' that the decoder portion of the auto-encoder is trying to match.

Can I use the PDist layer to implement these feature map reconstruction losses? I don't want to propagate derivatives for the feature map inputs from the encoder side of the SWWAE architecture, as they're the targets trying to be matched by the decoder side. I'm guessing that I should copy the DAG PDist class and just turn off or don't return the derivatives computed for the input target feature maps?

Any guidance would be greatly appreciated! Thanks in advance.

![swwae_architecture](https://cloud.githubusercontent.com/assets/27079732/24830782/940e148a-1c5a-11e7-884e-76e1dab01fef.png)

**Comment by hawescm1** [comment]
I tried my suggestion above, of copying the DAG PDist class and not returning the derivatives computed for the input target feature maps. This failed, as there is an error check in the Layer class that the number of derivatives returned by my modified PDist class (call it PDist2Ref) must match the number of inputs to the PDist2Ref layer... Does MatConvNet have an 'ignoreDerivatives' field that I could use, or could something like this be implemented?

**Comment by hawescm1** [comment]
I ultimately resolved this by not using the PDist layer. Instead I wrote my own dagnn.LpLoss layer and now I can train my SWWAE architecture using MCN's DAG.

**Comment by hawescm1** [comment]
Here's the LpLoss class I defined:

classdef LpLoss < dagnn.ElementWise
  properties
    opts = {};
    p = 2;
    ignoreAverage = false;
  end

  properties (Transient)
    average = 0
    numAveraged = 0
  end

  methods
    function outputs = forward(obj, inputs, params)
      sz = size(inputs{1});
      % We assume that inputs{1} is the estimate and inputs{2}
      % is the reference/target being estimated. We compute the
      % Lp-norm, without the pth-root:
      delta = inputs{1}(:) - inputs{2}(:);

      % Lp-norm without pth-root:
      if (obj.p == 1)
        % L1 norm: ||in_1 - in_2||_1^1
        outputs{1} = sum( abs(delta) ) / numel(inputs{1});

      elseif (obj.p == 2)
        % squared L2 norm: ||in_1 - in_2||_2^2
        outputs{1} = sum(delta.^2) / numel(inputs{1});

      else
        % pth power of Lp norm (for p != 1):  ||in_1 - in_2||_p^p
        outputs{1} = sum(abs(delta).^obj.p) / numel(inputs{1});
      end
      
      obj.accumulateAverage(inputs,outputs);
    end

    function [derInputs, derParams] = backward(obj, inputs, params, derOutputs)
      epsilon = 1e-6;
      sz = size(inputs{1});
      dzdy = derOutputs{1};
      
      % We assume that inputs{1} is the estimate and inputs{2}
      % is the reference trying to be estimated:
      delta = inputs{1} - inputs{2};
      
      if (obj.p == 1)
        % L1 norm:
        derInputs{1} = bsxfun(@times, dzdy, sign(delta));
        
      elseif (obj.p == 2)
        % squared L2 norm:
        derInputs{1} = bsxfun(@times, 2*dzdy, delta);
        
      elseif (obj.p < 1)
        % pth-power of Lp, for p < 1, is not a norm:
        derInputs{1} = bsxfun(@times, obj.p*dzdy, max(abs(delta),epsilon).^(obj.p-1) .* sign(delta));
        
      else
        % pth-power of Lp norm, for p > 1:
        derInputs{1} = bsxfun(@times, obj.p*dzdy, abs(delta).^(obj.p-1) .* sign(delta));
      end
      
      % Normalize. Unlike the forward pass, we do not normalize by
      % the full number of elements in inputs{1}, as cnn_train_dag()
      % scales the accumulated gradients by the batch size in the
      % backward pass. Thus, we do not scale here by the batch size:
      derInputs{1} = derInputs{1} / (sz(1) * sz(2) * sz(3));
      derInputs{2} = [];
      derParams = {};
    end

    function accumulateAverage(obj, inputs, outputs)
      % We want to override the default behavior of accumulateAverage as
      % defined in the base class 'dagnn.Loss'
      if obj.ignoreAverage, return; end;
      n = obj.numAveraged;
      m = n + numel(inputs{1});
      obj.average = bsxfun(@plus, n * obj.average, gather(outputs{1})) / m;
      obj.numAveraged = m;
    end
    
    function reset(obj)
      obj.average = 0;
      obj.numAveraged = 0;
    end

    function outputSizes = getOutputSizes(obj, inputSizes, paramSizes)
      outputSizes{1} = [1 1 1 inputSizes{1}(4)];
    end

    function rfs = getReceptiveFields(obj)
      % the receptive field depends on the dimension of the variables
      % which is not known until the network is run
      rfs(1,1).size   = [NaN NaN];
      rfs(1,1).stride = [NaN NaN];
      rfs(1,1).offset = [NaN NaN];
      rfs(2,1) = rfs(1,1);
    end

    function obj = LpLoss(varargin)
      obj.load(varargin);
    end
  end
end

**Comment by nicjac** [comment]
@hawescm1 this looks very interesting. Would you mind sharing your SWWAE implementation? Thanks!
